src/placeos-triggers/mapping.cr

```diff
@@ -0,0 +1,99 @@
+require "./state"
+
+module PlaceOS::Triggers
+  class Mapping
+    private getter mapping_lock = Mutex.new(protection: :reentrant)
+
+    private getter trigger_cache = {} of String => Model::Trigger
+    private getter trigger_state = Hash(String, Array(State)).new { |hash, key| hash[key] = [] of State }
+    private getter instances = {} of String => State
+
+    def with_instances(& : Hash(String, State) ->)
+      mapping_lock.synchronize do
+        yield self.instances
+      end
+    end
+
+    def with_cache(& : Hash(String, Model::Trigger) ->)
+      mapping_lock.synchronize do
+        yield self.trigger_cache
+      end
+    end
```

Comment:
Dropping direct access to these internals (which looks doable with some light refactoring) would go a _long_ way to improve safety.

Reply:
don't think `with_cache` is in use, and `with_instances` can be made simpler.